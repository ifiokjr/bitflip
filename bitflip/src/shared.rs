pub struct AppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
#[cfg(feature = "ssr")]
impl axum::response::IntoResponse for AppError {
	fn into_response(self) -> axum::response::Response {
		(
			http::StatusCode::INTERNAL_SERVER_ERROR,
			format!("App error: {}", self.0),
		)
			.into_response()
	}
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to
// turn them into `Result<_, AppError>`. That way you don't need to do that
// manually.
impl<E> From<E> for AppError
where
	E: Into<anyhow::Error>,
{
	fn from(err: E) -> Self {
		Self(err.into())
	}
}
