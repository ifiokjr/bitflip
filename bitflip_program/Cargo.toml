[package]
name = "bitflip_program"
version = "0.0.0"
edition = "2021"
description = "Created with Anchor"

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[features]
default = ["cpi"]
cpi = []
no-entrypoint = []
no-idl = []
no-log-ix-name = []
idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
serde = ["dep:serde", "dep:serde_with"]
client = ["serde", "no-entrypoint", "dep:typed-builder"]

[dependencies]
anchor-lang = { workspace = true, features = ["interface-instructions", "init-if-needed"] }
anchor-spl = { workspace = true }
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
typed-builder = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
test_utils = { workspace = true }
test_utils_solana = { workspace = true }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
