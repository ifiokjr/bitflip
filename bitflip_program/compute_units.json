{
	"instructions": {
		"config_initialize": {
			"compute_units": 17569,
			"rounded_compute_units": 20000,
			"description": "Initialize the config"
		},
		"config_update_authority": {
			"compute_units": 2326,
			"rounded_compute_units": 10000,
			"description": "Update the authority of the config"
		},
		"flip_bit_test": {
			"compute_units": 64832,
			"rounded_compute_units": 70000,
			"description": "Flip a single bit from `0` to `1`"
		},
		"game_initialize": {
			"compute_units": 12979,
			"rounded_compute_units": 20000,
			"description": "Initialize the game"
		},
		"game_refresh_signer": {
			"compute_units": 2232,
			"rounded_compute_units": 10000,
			"description": "Refresh the signer"
		},
		"token_group_initialize": {
			"compute_units": 47429,
			"rounded_compute_units": 50000,
			"description": "Initialize the token group"
		},
		"token_initialize:bit": {
			"compute_units": 63184,
			"rounded_compute_units": 70000,
			"description": "Initialize group token: Bit"
		},
		"token_initialize:gibibit": {
			"compute_units": 59746,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: GibiBit"
		},
		"token_initialize:kibibit": {
			"compute_units": 59745,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: KibiBit"
		},
		"token_initialize:mebibit": {
			"compute_units": 59742,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: MebiBit"
		}
	},
	"version": "0.0.0"
}
