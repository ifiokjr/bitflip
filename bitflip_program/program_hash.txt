ebc22d77bc8ddd23f326598673e9fe7aa99fcdce9572fb8559a8d9782ade4ab8
