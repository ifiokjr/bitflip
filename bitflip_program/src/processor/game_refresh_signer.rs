use steel::*;

use super::BitflipInstruction;
use crate::create_pda_game;
use crate::GameState;
use crate::ID;

pub fn process_game_refresh_signer(accounts: &[AccountInfo]) -> ProgramResult {
	let [access_signer_info, refresh_signer_info, game_info] = accounts else {
		return Err(ProgramError::InvalidArgument);
	};

	access_signer_info.is_signer()?;
	refresh_signer_info.is_signer()?.is_writable()?;
	game_info.is_type::<GameState>(&ID)?.is_writable()?;

	let game_state = game_info.as_account_mut::<GameState>(&ID)?;
	let game_key = create_pda_game(game_state.game_index, game_state.bump)?;

	if game_info.key.ne(&game_key) {
		return Err(ProgramError::InvalidSeeds);
	}

	game_state.access_signer = *access_signer_info.key;

	Ok(())
}

#[repr(C)]
#[derive(Clone, Copy, Debug, Pod, Zeroable)]
pub struct GameRefreshSigner {}

instruction!(BitflipInstruction, GameRefreshSigner);

#[cfg(test)]
mod tests {
	use assert2::check;

	use super::*;
	use crate::get_pda_game;
	use crate::leak;
	use crate::process_game_initialize;

	#[test_log::test]
	fn should_have_enough_accounts() -> anyhow::Result<()> {
		let accounts = create_account_infos();
		let result = process_game_initialize(&accounts[..2]);
		check!(result.unwrap_err() == ProgramError::NotEnoughAccountKeys);

		Ok(())
	}

	#[test_log::test]
	fn access_signer_should_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let access_signer_info = &mut accounts[0];
		access_signer_info.is_signer = false;

		let result = process_game_refresh_signer(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn refresh_signer_should_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let refresh_signer_info = &mut accounts[1];
		refresh_signer_info.is_signer = false;

		let result = process_game_refresh_signer(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn refresh_signer_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let refresh_signer_info = &mut accounts[1];
		refresh_signer_info.is_writable = false;

		let result = process_game_refresh_signer(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn game_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let game_info = &mut accounts[2];
		game_info.key = leak(Pubkey::new_unique());

		let result = process_game_refresh_signer(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn game_should_have_valid_owner() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let game_info = &mut accounts[2];
		game_info.owner = leak(Pubkey::new_unique());

		let result = process_game_refresh_signer(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidAccountOwner);

		Ok(())
	}

	#[test_log::test]
	fn game_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let game_info = &mut accounts[2];
		game_info.is_writable = false;

		let result = process_game_refresh_signer(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	fn create_account_infos<'info>() -> [AccountInfo<'info>; 3] {
		let access_signer_key = leak(Pubkey::new_unique());
		let access_signer_lamports = leak(0);
		let access_signer_data = leak(vec![]);
		let refresh_signer_key = leak(Pubkey::new_unique());
		let refresh_signer_lamports = leak(500_000);
		let refresh_signer_data = leak(vec![]);
		let game_index = 0;
		let game_key = leak(get_pda_game(game_index).0);
		let game_lamports = leak(0);
		let game_data = {
			let game_bump = get_pda_game(game_index).1;
			let mut data = vec![0u8; 8];
			data[0] = GameState::discriminator();
			data.append(
				&mut GameState::new(
					*access_signer_key,
					*refresh_signer_key,
					game_index,
					game_bump,
				)
				.to_bytes()
				.to_vec(),
			);
			leak(data)
		};

		let access_signer_info = AccountInfo::new(
			access_signer_key,
			true,
			false,
			access_signer_lamports,
			access_signer_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let refresh_signer_info = AccountInfo::new(
			refresh_signer_key,
			true,
			true,
			refresh_signer_lamports,
			refresh_signer_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let game_info = AccountInfo::new(
			game_key,
			false,
			true,
			game_lamports,
			game_data,
			&ID,
			false,
			u64::MAX,
		);

		[access_signer_info, refresh_signer_info, game_info]
	}
}
