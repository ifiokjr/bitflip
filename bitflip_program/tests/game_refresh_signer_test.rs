#![cfg(feature = "client")]

use std::future::Future;

use assert2::check;
use bitflip_program::GameState;
use bitflip_program::game_refresh_signer;
use bitflip_program::get_pda_game;
use shared::CreatedGameState;
use shared::ToRpcClient;
use shared::create_config_accounts;
use shared::create_game_state;
use shared::create_program_context_with_factory;
use shared::create_token_accounts;
use solana_sdk::signature::Keypair;
use solana_sdk::transaction::VersionedTransaction;
use steel::*;
use test_utils::create_insta_redaction;
use test_utils_solana::prelude::*;

mod shared;

#[test_log::test(tokio::test)]
async fn game_refresh_signer_test() -> anyhow::Result<()> {
	let created_game_state = create_game_state(0, 0, 0, 0);
	shared_game_refresh_signer_test(
		|| create_banks_client_rpc(&created_game_state),
		&created_game_state,
	)
	.await?;

	Ok(())
}

#[cfg(feature = "test_validator")]
#[test_log::test(tokio::test)]
async fn game_refresh_signer_test_validator() -> anyhow::Result<()> {
	let created_game_state = create_game_state(0, 0, 0, 0);
	let compute_units = shared_game_refresh_signer_test(
		|| create_validator_rpc(&created_game_state),
		&created_game_state,
	)
	.await?;
	let rounded_compute_units = bitflip_program::round_compute_units_up(compute_units);

	check!(rounded_compute_units == 10_000);
	insta::assert_snapshot!(format!("{rounded_compute_units} CU"));
	shared::save_compute_units("game_refresh_signer", compute_units, "Refresh the signer")?;
	Ok(())
}

async fn create_banks_client_rpc(
	CreatedGameState {
		ref game_state_account,
		ref refresh_signer,
		ref refresh_signer_account,
		..
	}: &CreatedGameState,
) -> anyhow::Result<impl ToRpcClient> {
	let provider = create_program_context_with_factory(|p| {
		let mut config_state_accounts = create_config_accounts();
		config_state_accounts.extend(create_token_accounts()?);

		for (key, account) in config_state_accounts {
			p.add_account(key, account.into());
		}

		let game = get_pda_game(0).0;

		p.add_account(game, game_state_account.clone().into());
		p.add_account(
			refresh_signer.pubkey(),
			refresh_signer_account.clone().into(),
		);

		Ok(())
	})
	.await?;

	Ok(provider)
}

#[cfg(feature = "test_validator")]
async fn create_validator_rpc(
	CreatedGameState {
		ref game_state_account,
		ref refresh_signer,
		ref refresh_signer_account,
		..
	}: &CreatedGameState,
) -> anyhow::Result<impl ToRpcClient> {
	let mut accounts = create_config_accounts();
	let game = get_pda_game(0).0;
	accounts.insert(game, game_state_account.clone());
	accounts.insert(refresh_signer.pubkey(), refresh_signer_account.clone());

	let runner = shared::create_runner_with_accounts(accounts).await;

	Ok(runner)
}

async fn shared_game_refresh_signer_test<
	T: ToRpcClient,
	Fut: Future<Output = anyhow::Result<T>>,
	Create: FnOnce() -> Fut,
>(
	create_provider: Create,
	CreatedGameState {
		refresh_signer: ref refresh_signer_keypair,
		..
	}: &CreatedGameState,
) -> anyhow::Result<u64> {
	let provider = create_provider().await?;
	let rpc = provider.to_rpc();
	let access_signer_keypair = Keypair::new();
	let access_signer = access_signer_keypair.pubkey();
	let refresh_signer = refresh_signer_keypair.pubkey();
	let game_index = 0;
	let game = get_pda_game(game_index).0;

	// Create transaction
	let recent_blockhash = rpc.get_latest_blockhash().await?;
	let ix = game_refresh_signer(&access_signer, &refresh_signer, game_index);
	let mut transaction =
		VersionedTransaction::new_unsigned_v0(&refresh_signer, &[ix], &[], recent_blockhash)?;

	let simulation = rpc.simulate_transaction(&transaction).await?;
	log::info!("simulation: {simulation:#?}");
	let compute_units = simulation.value.units_consumed.unwrap();

	// Sign and send transaction
	transaction
		.try_sign(&[&access_signer_keypair], None)?
		.try_sign(&[refresh_signer_keypair], None)?;

	let signature = rpc.send_and_confirm_transaction(&transaction).await?;
	rpc.confirm_transaction(&signature).await?;

	// Verify the game state was updated
	let game_state_data = rpc.get_account_data(&game).await?;
	let game_state_account = GameState::try_from_bytes(&game_state_data)?;
	let access_signer_redaction = create_insta_redaction(access_signer, "access_signer:pubkey");
	let refresh_signer_redaction = create_insta_redaction(refresh_signer, "refresh_signer:pubkey");
	insta::assert_compact_json_snapshot!(game_state_account,{
		".accessSigner" => insta::dynamic_redaction(access_signer_redaction),
		".refreshSigner" => insta::dynamic_redaction(refresh_signer_redaction),

	}, @r#"
 {
   "refreshSigner": "[refresh_signer:pubkey]",
   "accessSigner": "[access_signer:pubkey]",
   "accessExpiry": 0,
   "startTime": 0,
   "gameIndex": 0,
   "sectionIndex": 0,
   "bump": 253
 }
 "#);

	Ok(compute_units)
}
