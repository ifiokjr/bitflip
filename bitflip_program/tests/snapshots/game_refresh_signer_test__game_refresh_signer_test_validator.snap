---
source: bitflip_program/tests/game_refresh_signer_test.rs
expression: "format!(\"{rounded_compute_units} CU\")"
snapshot_kind: text
---
10000 CU
